loki/transformations/build_system/plan.py

```diff
@@ -96,36 +96,72 @@ def plan_file(self, sourcefile, **kwargs):
 
         debug(f'Planning:: {item.name} (role={item.role}, mode={item.mode})')
 
+        key = item.lib
         if newsource not in self.sources_to_append:
             if source_exists:
-                self.sources_to_transform += [sourcepath]
+                self.sources_to_transform.setdefault(key,[]).append(sourcepath)
             if item.replicate:
                 # Add new source file next to the old one
-                self.sources_to_append += [newsource]
+                self.sources_to_append.setdefault(key,[]).append(newsource)
             else:
                 # Replace old source file to avoid ghosting
-                self.sources_to_append += [newsource]
+                self.sources_to_append.setdefault(key,[]).append(newsource)
                 if source_exists:
                     if self.rootpath is not None:
-                        self.sources_to_remove += [sourcepath]
+                        self.sources_to_remove.setdefault(key,[]).append(sourcepath)
                     else:
                         # NB: we use the item path directly here instead of resolving it,
                         #     to stay compatible with what has been provided on the CLI.
                         #     This is because the build system will likely use this path
                         #     internally to identify the original file, and if the paths
                         #     don't match the removal of source files from the target fails.
-                        self.sources_to_remove += [item.path]
+                        self.sources_to_remove.setdefault(key,[]).append(item.path)
 
-    def write_plan(self, filepath):
+    def _write_plan(self, filepath):
         """
-        Write the CMake plan file to :data:`filepath`
+        Write the key/target/library independent part of CMake plan file to :data:`filepath`
         """
+        sources_to_transform = []
+        sources_to_append = []
+        sources_to_remove = []
+        for _, vals in self.sources_to_transform.items():
+            sources_to_transform.extend(vals)
+        for _, vals in self.sources_to_append.items():
+            sources_to_append.extend(vals)
+        for _, vals in self.sources_to_remove.items():
+            sources_to_remove.extend(vals)
```

Comment:
Instead of `items`, the canonical pattern would be to use `values`, which you can then shorten as:

```suggestion
        sources_to_transform = [s for sources in self.sources_to_transform.values() for s in sources]
        sources_to_append = [s for sources in self.sources_to_append.values() for s in sources]
        sources_to_remove = [s for sources in self.sources_to_remove.values() for s in sources]
```